### ISSUE TITLE
Add new brace-style option supporting long multi-line declarations

### ISSUE THREAD

**Issue by JakubKoralewski** [eslint/eslint#12250]
This is idea may be novel since, class declarations are usually quite short but when you think about the declaration in TypeScript being able to have both the `export`, `default`, `class`, `VeryLongName`, `extends`, `OtherLongName`, `implements`, `OtherLongName` then you can see how the maximum line length can be reached leading to the declaration spanning multiple lines breaking the `max-len` rule.

Then you could argue I should post this on TSLint, which I originally planned to do, but as I just found out TSLint is on a roadmap to be completely replaced by ESLint for Typescript by 2019. It makes sense to add a feature to something that's going to be ultimately merged with TSLint.

Actually same goes for JavaScript with lots of arguments that are wrapped between multiple lines, e.g.:
```ts
function getFavor(
	favor: TheFavor,
	permission: Element,
	type: string,
): Element {
	const theFavor = favor.new();
        return theFavor;
}
````

**What rule do you want to change?**

https://eslint.org/docs/rules/brace-style

**Does this change cause the rule to produce more or fewer warnings?**

If set then more, but that's why its set -- to throw warnings where it didn't before?

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option.

**Please provide some example code that this change will affect:**

```ts
export default class ReferencableToUID extends ReferencableOrOtherVeryLongNameToCauseMaxLenToThrowError {
```

**What does the rule currently do for this code?**

`max-len` throws error:
`This line has a length of 105. Maximum allowed is 80.`

So then I go and move the definition to two lines:

```ts
export default class ReferencableToUID 
        extends ReferencableOrOtherVeryLongNameToCauseMaxLenToThrowError {
```

Which removes `max-len`, but then the problem is that I don't like how the code looks now. The horizontal level of the brace may (depending on the declaration length) be very close to the function declarations below making the distinction between the declaration and the (e.g.) class body hard to see.

```ts
export default class ReferencableToUID 
  extends ReferencableOrOtherVeryLongNameToCauseMaxLenToThrowError {
	getReferencedObject(xml, type) {
		const uRef = this.element.getAttribute('ObjectURef');
		let found = xml.querySelector(getQueryWithURef(uRef, type));
		if (!found) { \\ (...)
```

**What will the rule do after it's changed?**

Allow to have either of the two following behaviors:

1. Add new brace-style that allows to have the brace on a new-line after a multi-line declaration, effectively changing the style to "allman":
```ts
export default class ReferencableToUID 
  extends ReferencableOrOtherVeryLongNameToCauseMaxLenToThrowError 
{
	getReferencedObject(xml, type) {
		const uRef = this.element.getAttribute('ObjectURef');
		let found = xml.querySelector(getQueryWithURef(uRef, type));
		if (!found) { \\ (...)

```

2. Padding between multi-line declaration and first property / function:
```ts
export default class ReferencableToUID 
  extends ReferencableOrOtherVeryLongNameToCauseMaxLenToThrowError {
	
        getReferencedObject(xml, type) {
		const uRef = this.element.getAttribute('ObjectURef');
		let found = xml.querySelector(getQueryWithURef(uRef, type));
		if (!found) { \\ (...)
```

I'm not sure about the naming of the new style, but ultimately the behavior is something like the "allman" brace-style but only when the declaration spans multiple lines and defaults to "Stroustrup".

**Are you willing to submit a pull request to implement this change?**

No.

**Comment by kaicataldo** [comment]
I don't think this is a common enough style to include in core, but seems like a great candidate for a plugin or custom rule.
